### ISSUE TITLE
error: [Errno 32] Broken pipe

### ISSUE THREAD

**Issue by CohibAA** [JoinMarket-Org/joinmarket#194]
This happens sporadically with yield-generator.py, showing in the terminal logs.  This doesn't _seem_ to cause any obvious issues, other than the message.

>   127.0.0.1 - - [15/Aug/2015 xx:xx:xx] "HEAD /walletnotify?redactedtxid HTTP/1.1" 200 -
\----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 35044)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 295, in \_handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 321, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
File "/home/cohibaa/joinmarket/lib/blockchaininterface.py", line 277, in \__init__
    BaseHTTPServer.BaseHTTPRequestHandler.\__init__(self, request, client_address, base_server)
  File "/usr/lib/python2.7/SocketServer.py", line 651, in \__init__
    self.finish()
  File "/usr/lib/python2.7/SocketServer.py", line 704, in finish
    self.wfile.flush()
  File "/usr/lib/python2.7/socket.py", line 303, in flush
    self._sock.sendall(view[write_offset:write_offset+buffer_size])
> error: [Errno 32] Broken pipe

> ----------------------------------------

I think it happens during [these lines](https://github.com/chris-belcher/joinmarket/blob/master/lib/blockchaininterface.py#L329-#L330), as everything works normally up to and after that point.

**Comment by CohibAA** [comment]
By sporadically, I mean that sometimes [end_headers()](https://github.com/chris-belcher/joinmarket/blob/master/lib/blockchaininterface.py#L329-#L330) will execute fully, without a broken pipe.

**Comment by tailsjoin** [comment]
4/4 of my makers transactions today had this error. As was stated, maker recovers with no issue.

**Comment by CohibAA** [comment]
For the record, you won't see this in the maker's log file, it only prints to the terminal.  I captured the errors using something like

    screen -L python yield-generator.py wallet.json

where -L saves everything to screenlog.0.  I've added some extra debug lines to my maker to monitor this, and I am confident that it is failing on [line #330](https://github.com/chris-belcher/joinmarket/blob/master/lib/blockchaininterface.py#L330) for some reason.  Line 328 is always successful.  

Again, this does not have any obvious affect on the bot, including order amounts, utxos available, etc.  ELI5 what L330 does, anyone?

**Comment by AdamISZ** [comment]
@CohibAA  I'm looking at it, and I follow your logic, only to say: no need to think about bitcoin rpc, it's curl's behaviour we're interested in here, right (specifically curl -I). 

I'm thinking: can I reliably reproduce the broken pipe? If so, I can wireshark it at worst :)

**Comment by AdamISZ** [comment]
A possible theory: the walletnotify calls curl with connect-timeout 1 second. within that curl call, which goes to the do_HEAD() call, we don't send the end_headers() until *after* we call curl --connect-timeout 1 again, on line 328. This second call is to a bot listening on 62603, but there is no bot (if you're only running one). So that call will time out, and will create a delay of 1 second. This means that the first curl (coming directly from bitcoin) has given up waiting (it took more than 1 second) and closed the connection. So by the time we call end_headers(), the socket is closed and you get the broken pipe.

This doesn't explain why you would only get the broken pipe occasionally though. If indeed it makes sense at all...

**Comment by CohibAA** [comment]
@AdamISZ I also *thought* that made sense.  However, I have already tried changing my walletnotify timeout in bitcoin.conf to 5 seconds (and restarted the node) and the elusive broken pipes persist.  It was actually one of the first things I tried, before even posting this issue.  Looking at the debug maker logs, it all happens within 1 second anyway, afaict (my logs doesn't display fractions of seconds), but certainly well under 5 seconds.

Also, commenting out and not sending end_headers() does NOT stop the broken pipe, yet the send_response(200) seems to have executed without issue.  Have you been able to replicate at all (even if not consistently)?  It happens to my maker on virtually every transaction at least once, until after the tx is in a block.  Maybe it's coming from somewhere else?

**Comment by tailsjoin** [comment]
It's happening for me all the time too, which leads me to wonder if this is somehow connected to tor.

**Comment by tailsjoin** [comment]
Going off my assumption about this being tor related, I did a little research. I think this [tor ticket](https://trac.torproject.org/projects/tor/ticket/9998) might have something to do with it. Therefore I changed `localhost` from wallet/alertnotify in bitcoin.conf to `127.0.0.1` and then also in lib/blockchaininterface.py here: https://github.com/chris-belcher/joinmarket/blob/master/lib/blockchaininterface.py#L326. Since then I haven't seen this in my logs.

**Comment by CohibAA** [comment]
@tailsjoin Good find, unfortunately, I am still seeing them even after changing references to localhost to 127.0.0.1 - are you still not seeing any?

I might have another idea.

**Comment by tailsjoin** [comment]
My tests proved inconclusive. I ran two bots, one with the fix and one without. I did 10 txs using both of them and saw the error only once on the bot without the fix. Since then I haven't had any txs and I haven't tested any more.

**Comment by CohibAA** [comment]
Can anyone confirm, has this error been seen in standard out when **not** using tor?  I have been running some bots over clearnet and have not seen this (yet?).  I'd like to rule out tor, but I'm starting to think that might be it.  Unfortunately, I've thought a lot of wrong things about this annoying :bug:.

**Comment by dooglus** [comment]
As mentioned in #219, I see this regularly and don't use Tor.

**Comment by meeDamian** [comment]
I'm afraid I need to revive this one.

I'm getting below ~75% of the time:

```
127.0.0.1 - - [30/Dec/2016 06:35:56] "HEAD /walletnotify?<txid> HTTP/1.1" 200 -
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 45662)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 290, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 318, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/home/mee/repos/joinmarket/joinmarket/blockchaininterface.py", line 440, in __init__
    self, request, client_address, base_server)
  File "/usr/lib/python2.7/SocketServer.py", line 654, in __init__
    self.finish()
  File "/usr/lib/python2.7/SocketServer.py", line 713, in finish
    self.wfile.close()
  File "/usr/lib/python2.7/socket.py", line 283, in close
    self.flush()
  File "/usr/lib/python2.7/socket.py", line 307, in flush
    self._sock.sendall(view[write_offset:write_offset+buffer_size])
error: [Errno 32] Broken pipe
```

I'm using `127.0.0.1` everywhere possible and not using tor.
